library/dnet.c

```diff
@@ -393,6 +406,375 @@ static int dnet_iterator_callback_send(void *priv, void *data, uint64_t dsize)
+		 * local record, but also do not stop iterator.
+		 *
+		 * Setting @write_error forces iterator to stop.
+		 */
+		if (err && !send->write_error && (err != -EBADFD))
+			send->write_error = err;
+
+		if (atomic_dec_and_test(&wp->refcnt)) {
+			if (send->iflags & DNET_IFLAGS_MOVE) {
+				if (!err) {
+					struct dnet_io_req *r;
+					struct dnet_cmd *lc;
+					struct dnet_io_attr *io;
+					size_t cmd_size = sizeof(struct dnet_io_req) +
+							sizeof(struct dnet_cmd) +
+							sizeof(struct dnet_io_attr);
+
+					r = malloc(cmd_size);
+					if (!r) {
+						err = -ENOMEM;
+						// we could update wp->data here, which is dnet_iterator_response
+						// but we do not really care about local errors, for example
+						// remove error is not handled too
+						goto err_out_send;
+					}
+
+					memset(r, 0, cmd_size);
+
+					r->header = r + 1;
+					r->hsize = sizeof(struct dnet_cmd);
+					r->data = r->header + sizeof(struct dnet_cmd);
+					r->dsize = sizeof(struct dnet_io_attr);
+					r->st = dnet_state_get(st);
+
+					lc = r->header;
+					dnet_setup_id(&lc->id, send->cmd.id.group_id, cmd->id.id);
+					lc->cmd = DNET_CMD_DEL;
+					lc->backend_id = -1;
+					lc->trace_id = cmd->trace_id;
+					lc->flags = DNET_FLAGS_NOLOCK;
+					lc->size = sizeof(struct dnet_io_attr);
+
+					io = r->data;
+					io->flags = DNET_IO_FLAGS_SKIP_SENDING;
+					memcpy(io->id, lc->id.id, DNET_ID_SIZE);
+					memcpy(io->parent, lc->id.id, DNET_ID_SIZE);
+					dnet_convert_io_attr(io);
+
+					dnet_schedule_io(st->n, r);
+				}
+			}
+
+err_out_send:
+			err = dnet_send_reply(send->state, &send->cmd, wp->data, wp->dsize, 1);
+			if (err && !send->write_error)
+				send->write_error = err;
+
+			struct dnet_iterator_response *re = (struct dnet_iterator_response *)wp->data;
+			dnet_convert_iterator_response(re);
+			dnet_log(st->n, DNET_LOG_NOTICE, "%s: %s: sending response to client: %s, "
+					"user_flags: %llx, ts: %lld.%09lld, "
+					"status: %d, size: %lld, iterated_keys: %lld/%lld, write_error: %d",
+					__func__,
+					dnet_dump_id(&send->cmd.id), dnet_dump_id_str(re->key.id),
+					(unsigned long long)re->user_flags,
+					(unsigned long long)re->timestamp.tsec, (unsigned long long)re->timestamp.tnsec,
+					re->status, (unsigned long long)re->size,
+					(unsigned long long)re->iterated_keys, (unsigned long long)re->total_keys,
+					send->write_error);
+
+			atomic_dec(&send->writes_pending);
+			pthread_cond_broadcast(&send->write_wait);
+
+			dnet_server_send_put(send);
+			free(wp);
+			return err;
+		}
+	}
+
+	return err;
+}
+
+struct dnet_server_send_ctl *dnet_server_send_alloc(void *state, struct dnet_cmd *cmd, uint64_t iflags,
+		int *groups, int group_num)
+{
+	int err;
+	struct dnet_net_state *st = state;
+	struct dnet_server_send_ctl *ctl;
+
+	dnet_state_get(st);
+
+	ctl = malloc(sizeof(struct dnet_server_send_ctl) + sizeof(int) * group_num);
+	if (!ctl) {
+		err = -ENOMEM;
```

Comment:
This is a very bad idea to remove error assignment just because it is not used.
Eventually we may want to print it or return and may forget that it was not assigned in some error path.